File: src/main/java/com/atsuishio/superbwarfare/perk/damage/VorpalWeapon.java
package com.atsuishio.superbwarfare.perk.damage;

import com.atsuishio.superbwarfare.data.gun.GunProp;
import com.atsuishio.superbwarfare.init.ModDamageTypes;
import com.atsuishio.superbwarfare.perk.Perk;
import com.atsuishio.superbwarfare.tools.DamageTypeTool;
import net.minecraft.world.entity.LivingEntity;

public class VorpalWeapon extends Perk {

    public VorpalWeapon() {
        super("vorpal_weapon", Perk.Type.DAMAGE);
        appendModification(GunProp.DAMAGE, ((data, value, target, source) -> {
            if (source == null) return value;
            if ((DamageTypeTool.isGunDamage(source) || source.is(ModDamageTypes.PROJECTILE_BOOM)) && target instanceof LivingEntity living && living.getHealth() >= 100.0f) {
                return value + living.getHealth() * 0.00002 * Math.pow(data.perk.getLevel(this), 2);
            }
            return value;
        }));
    }
}

File: src/main/resources/data/superbwarfare/sbw/perks/vorpal_weapon.js
// VORPAL_WEAPON: 对高血量目标额外伤害
function getModifiedDamage(damage, target, level, perkTag, sourceProxy) {
    if (sourceProxy && sourceProxy.isGunDamage() && target.health >= 100) {
        return damage + target.health * 0.00002 * Math.pow(level, 2)
    }
    return damage
}
